### PR TITLE
CNKAF-739: Add Rest endpoint for list all partition reassignments

### PR DESCRIPTION
**Pull request by agg111**
### What
Adds rest endpoint to fetch all ongoing partition reassignments - 
`GET /v3/clusters/<clusterId>/topics/-/partitions/-/reassignments`

### Testing
- Unit tests for resource and manager class.
- Integration test to verify parameters like adding_replicas for on-going reassignments.

### References 
One-pager - https://confluentinc.atlassian.net/wiki/spaces/CNKAF/pages/1276742825/REST+API+for+partition+reassignment
Jira - https://confluentinc.atlassian.net/browse/CNKAF-739

### CODE REVIEW

File: kafka-rest/src/test/java/io/confluent/kafkarest/integration/v3/ListAllReassignmentsActionIntegrationTest.java [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+
+  private Map<TopicPartition, Optional<NewPartitionReassignment>> createReassignment(
+      List<Integer> replicaIds) {
+    Map<TopicPartition, Optional<NewPartitionReassignment>> reassignmentMap = new HashMap<>();
+    for (int i = 0; i < 100; i++) {
+      reassignmentMap.put(new TopicPartition(TOPIC_NAME, i),
+          Optional.of(new NewPartitionReassignment(replicaIds)));
+    }
+    return reassignmentMap;
+  }
```
Comment (agg111) [inline kafka-rest/src/test/java/io/confluent/kafkarest/integration/v3/ListAllReassignmentsActionIntegrationTest.java:88]: As these methods will be used in several places, will create a util class in a follow up PR.

Comment (rigelbm) [reply kafka-rest/src/test/java/io/confluent/kafkarest/integration/v3/ListAllReassignmentsActionIntegrationTest.java:90]: You can add them to `ClusterHarness`.

**Comment by agg111** [comment]
Thanks @rigelbm , I have addressed comments. Will merge after build passes.

File: kafka-rest/src/main/java/io/confluent/kafkarest/entities/v3/ListReassignmentsResponse.java [hunk 2]
```diff
@@ -0,0 +1,77 @@
+/*
```
Comment (rigelbm) [inline kafka-rest/src/main/java/io/confluent/kafkarest/entities/v3/ListReassignmentsResponse.java:1]: I believe this file is left-over from the merge. This got renamed ListAllReassignmentsResponse. Please, delete.

File: kafka-rest/src/test/java/io/confluent/kafkarest/integration/v3/ListAllReassignmentsActionIntegrationTest.java [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+              .targetReplicas());
+    }
+  }
+
+  @Test
+  public void listAllReassignments_nonExistingCluster_returnsNotFound() throws Exception {
+
+    Response response = request("/v3/clusters/foobar/topics/-/partitions"
+        + "/-/reassignments")
+        .accept(Versions.JSON_API)
```
Comment (rigelbm) [inline kafka-rest/src/test/java/io/confluent/kafkarest/integration/v3/ListAllReassignmentsActionIntegrationTest.java:68]: Change to `MediaType.APPLICATION_JSON`.

File: kafka-rest/src/test/java/io/confluent/kafkarest/integration/v3/ListAllReassignmentsActionIntegrationTest.java [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+    String clusterId = getClusterId();
+
+    Map<TopicPartition, Optional<NewPartitionReassignment>> reassignmentMap =
+        createReassignment(Arrays.asList(3, 4, 5), TOPIC_NAME, 100);
+
+    alterPartitionReassignment(reassignmentMap);
+
+    Response response =
+        request("/v3/clusters/" + clusterId + "/topics/-/partitions/-/reassignments")
+            .accept(Versions.JSON_API)
```
Comment (rigelbm) [inline kafka-rest/src/test/java/io/confluent/kafkarest/integration/v3/ListAllReassignmentsActionIntegrationTest.java:48]: Change to `MediaType.APPLICATION_JSON`.

File: kafka-rest/src/test/java/io/confluent/kafkarest/integration/v3/ListAllReassignmentsActionIntegrationTest.java [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+          data.getAddingReplicas(),
+          reassignmentMap.get(new TopicPartition(TOPIC_NAME, data.getPartitionId()))
+              .get()
+              .targetReplicas());
+    }
+  }
+
+  @Test
+  public void listAllReassignments_nonExistingCluster_returnsNotFound() throws Exception {
+
```
Comment (rigelbm) [inline kafka-rest/src/test/java/io/confluent/kafkarest/integration/v3/ListAllReassignmentsActionIntegrationTest.java:65]: Remove new line.

File: kafka-rest/src/test/java/io/confluent/kafkarest/integration/v3/ListAllReassignmentsActionIntegrationTest.java [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+import java.util.Map;
+import java.util.Optional;
+import javax.ws.rs.core.Response;
+import javax.ws.rs.core.Response.Status;
+import org.apache.kafka.clients.admin.NewPartitionReassignment;
+import org.apache.kafka.common.TopicPartition;
+import org.junit.Before;
+import org.junit.Test;
+
+public class ListAllReassignmentsActionIntegrationTest extends ClusterTestHarness {
```
Comment (rigelbm) [inline kafka-rest/src/test/java/io/confluent/kafkarest/integration/v3/ListAllReassignmentsActionIntegrationTest.java:22]: `RunWith(Junit4.class)`

File: kafka-rest/src/test/java/io/confluent/kafkarest/controllers/ReassignmentManagerImplTest.java [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+    expect(clusterManager.getCluster(CLUSTER_ID)).andReturn(completedFuture(Optional.empty()));
+    replay(clusterManager);
+
+    try {
+      reassignmentManager.listReassignments(CLUSTER_ID).get();
+      fail();
+    } catch (ExecutionException e) {
+      assertEquals(NotFoundException.class, e.getCause().getClass());
+    }
+
```
Comment (rigelbm) [inline kafka-rest/src/test/java/io/confluent/kafkarest/controllers/ReassignmentManagerImplTest.java:171]: Remove new line.

File: kafka-rest/src/test/java/io/confluent/kafkarest/controllers/ReassignmentManagerImplTest.java [hunk 8]
```diff
@@ -0,0 +1,173 @@
...
+import org.apache.kafka.common.KafkaFuture;
+import org.apache.kafka.common.Node;
+import org.apache.kafka.common.TopicPartition;
+import org.easymock.EasyMockRule;
+import org.easymock.Mock;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+
+public class ReassignmentManagerImplTest {
```
Comment (rigelbm) [inline kafka-rest/src/test/java/io/confluent/kafkarest/controllers/ReassignmentManagerImplTest.java:49]: `RunWith(Junit4.class)`

File: kafka-rest/src/main/java/io/confluent/kafkarest/entities/Reassignment.java [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+  Reassignment() {
+  }
+
+  public abstract String getClusterId();
+
+  public abstract String getTopicName();
+
+  public abstract int getPartitionId();
+
+  public abstract List<Integer> getReplicas();
```
Comment (rigelbm) [inline kafka-rest/src/main/java/io/confluent/kafkarest/entities/Reassignment.java:34]: This field is not useful for us anymore. We use a relationship to replicas in the response.

File: kafka-rest/src/main/java/io/confluent/kafkarest/controllers/ReassignmentManagerImpl.java [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+      Entry<TopicPartition, PartitionReassignment> reassignment) {
+    return Reassignment.create(
+        clusterId,
+        reassignment.getKey().topic(),
+        reassignment.getKey().partition(),
+        reassignment.getValue().replicas(),
+        reassignment.getValue().addingReplicas(),
+        reassignment.getValue().removingReplicas());
+  }
+
```
Comment (rigelbm) [inline kafka-rest/src/main/java/io/confluent/kafkarest/controllers/ReassignmentManagerImpl.java:79]: Remove new line.

File: kafka-rest/src/main/java/io/confluent/kafkarest/controllers/ReassignmentManagerImpl.java [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+import java.util.List;
+import java.util.Map.Entry;
+import java.util.concurrent.CompletableFuture;
+import java.util.stream.Collectors;
+import javax.inject.Inject;
+import org.apache.kafka.clients.admin.Admin;
+import org.apache.kafka.clients.admin.PartitionReassignment;
+import org.apache.kafka.common.TopicPartition;
+
+public class ReassignmentManagerImpl implements ReassignmentManager {
```
Comment (rigelbm) [inline kafka-rest/src/main/java/io/confluent/kafkarest/controllers/ReassignmentManagerImpl.java:34]: package-private final

File: kafka-rest/src/main/java/io/confluent/kafkarest/controllers/ReassignmentManager.java [hunk 12]
```diff
@@ -0,0 +1,32 @@
...
+import java.util.concurrent.CompletableFuture;
+
+public interface ReassignmentManager {
+
+  /**
+   * Returns a list of Kafka {@link Reassignment Reassignments} for all Kafka {@link Topic Topics}
+   * for a Kafka {@link Cluster} with the given {@code clusterId}.
+   */
+  CompletableFuture<List<Reassignment>> listReassignments(String clusterId);
+
```
Comment (rigelbm) [inline kafka-rest/src/main/java/io/confluent/kafkarest/controllers/ReassignmentManager.java:31]: Remove new line.

File: kafka-rest/src/main/java/io/confluent/kafkarest/entities/v3/ListReassignmentsResponse.java (hunk 2 above)
Comment (agg111) [reply kafka-rest/src/main/java/io/confluent/kafkarest/entities/v3/ListReassignmentsResponse.java]: removed.

File: kafka-rest/src/test/java/io/confluent/kafkarest/integration/v3/ListAllReassignmentsActionIntegrationTest.java (hunk 6 above)
Comment (agg111) [reply kafka-rest/src/test/java/io/confluent/kafkarest/integration/v3/ListAllReassignmentsActionIntegrationTest.java:23]: Added it, didn't usually come across the annotation on integration test class .

File: kafka-rest/src/test/java/io/confluent/kafkarest/controllers/ReassignmentManagerImplTest.java (hunk 7 above)
Comment (agg111) [reply kafka-rest/src/test/java/io/confluent/kafkarest/controllers/ReassignmentManagerImplTest.java:174]: removed.

File: kafka-rest/src/test/java/io/confluent/kafkarest/controllers/ReassignmentManagerImplTest.java (hunk 8 above)
Comment (agg111) [reply kafka-rest/src/test/java/io/confluent/kafkarest/controllers/ReassignmentManagerImplTest.java:52]: done.

File: kafka-rest/src/main/java/io/confluent/kafkarest/entities/Reassignment.java (hunk 9 above)
Comment (agg111) [reply kafka-rest/src/main/java/io/confluent/kafkarest/entities/Reassignment.java]: Nice catch, will remove.

File: kafka-rest/src/main/java/io/confluent/kafkarest/controllers/ReassignmentManager.java (hunk 12 above)
Comment (agg111) [reply kafka-rest/src/main/java/io/confluent/kafkarest/controllers/ReassignmentManager.java]: removed.

**Pull request merged**